### PR TITLE
New time-related placeholders till New Day

### LINKED ISSUE #5849: Suggestion: Time until new day placeholder
### **Please explain your feature request to the best of your abilities:**

Add 3 placeholders that returns the time until next towny day. The same information returned from /towny time

Using 3 separated and non formatted values ensures that it can be easily formatted on demand. 

%townyadvanced_world_time_until_new_day_hours% 
%townyadvanced_world_time_until_new_day_minutes%
%townyadvanced_world_time_until_new_day_seconds%

### PR DESCRIPTION
**Pull request by LoJoSho**
#### Description: 

Adds new placeholders relating to time until a new day, which can be useful for linking up with other plugins, such as DeluxeMenus, and having a different item depending on how much time is left in the towns day, or a warning if there is less than x amount of time left in a day. 

Added Placeholders:
`%townyadvanced_time_until_new_day_hours_raw% - Display the raw number of hours until new day.`
`%townyadvanced_time_until_new_day_minutes_raw% - Display the raw number of minutesuntil new day.`
`%townyadvanced_time_until_new_day_seconds_raw% - Display the raw number of seconds until new day.`
`%townyadvanced_time_until_new_day_formatted% - Display the amount of time until a new day, formatted and translated.`
`%townyadvanced_time_until_new_day_hours_formatted% - Display the amount of hours until a new day, formatted and translated.`
`%townyadvanced_time_until_new_day_minutes_formatted% - Display the amount of minutes until a new day, formatted and translated.`
`%townyadvanced_time_until_new_day_seconds_formatted% - Display the amount of seconds until a new day, formatted and translated.`

____
#### New Nodes/Commands/ConfigOptions: 

N/A
____
#### Relevant Towny Issue ticket:

Closes #5849 
____
- [x] I have tested this pull request for defects on a server. 

By making this pull request, I represent that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the TownyAdvanced organization has the copyright to use and modify my contribution under the Towny [License](https://github.com/LlmDl/Towny/blob/master/LICENSE.md) for perpetuity.

### CODE REVIEW

File: src/com/palmergames/util/TimeMgmt.java [hunk 1]
```diff
@@ -109,4 +109,30 @@ public static String getFormattedTimeValue(double timeMillis) {
...
+		out = h + Translation.of("msg_hours");
+		return out;
+	}
+	// Returns translation of "msg_minutes" formatted, ex: "737 minutes"
+	public static String formatCountdownTimeMinutes(long l) {
+
+		String out = "";
+		int m = (int) (l / 60.0);
+		out += (out.length() > 0 ? ", " : "") + m + Translation.of("msg_minutes");
+		return out;
```
Comment (Warriorrrr) [inline src/com/palmergames/util/TimeMgmt.java:128]:
```suggestion
		return Duration.ofSeconds(l).toMinutes() % 60 + Translation.of("msg_minutes");
```

File: src/com/palmergames/util/TimeMgmt.java [hunk 2]
```diff
@@ -109,4 +109,30 @@ public static String getFormattedTimeValue(double timeMillis) {
...
+	}
+	// Returns translation of "msg_seconds" formatted, ex: "44248 seconds"
+	public static String formatCountdownTimeSeconds(long l) {
+
+		String out = "";
+		out = l + Translation.of("msg_seconds");
+		return out;
```
Comment (Warriorrrr) [inline src/com/palmergames/util/TimeMgmt.java:135]:
```suggestion
		return l % 60 + Translation.of("msg_seconds");
```

**Review by LlmDl — CHANGES_REQUESTED** [review]
I haven't found much that warrior didn't spot already, I think it's worth noting that `TownyTimerHandler.townyTime()` is going to be calling 

```java
	public static Long townyTime() {

		long secondsInDay = TownySettings.getDayInterval();

		// Get Calendar instance
		Calendar now = Calendar.getInstance();

		// Get current TimeZone
		TimeZone timeZone = now.getTimeZone();
		
		// Get current system time in milliseconds
		long timeMilli = System.currentTimeMillis();
		
		// Calculate the TimeZone specific offset (including DST)
		int timeOffset = timeZone.getOffset(timeMilli)/1000;

		return Math.floorMod(secondsInDay + (TownySettings.getNewDayTime() - ((timeMilli/1000) % secondsInDay) - timeOffset), secondsInDay);
	}
```
on a near-constant basis when this is used in a tab/scoreboard plugin and it should probably call upon a cached result instead of calculating over and over again. I do realize this would mean that the seconds placeholder is inaccurate.

**Review by Warriorrrr — APPROVED** [review]
Looks good, thank you for the first time pr!

**Review by LlmDl — APPROVED** [review]

**Comment by LoJoSho** [comment]
> I haven't found much that warrior didn't spot already, I think it's worth noting that `TownyTimerHandler.townyTime()` is going to be calling
> 
> ```java
> 	public static Long townyTime() {
> 
> 		long secondsInDay = TownySettings.getDayInterval();
> 
> 		// Get Calendar instance
> 		Calendar now = Calendar.getInstance();
> 
> 		// Get current TimeZone
> 		TimeZone timeZone = now.getTimeZone();
> 		
> 		// Get current system time in milliseconds
> 		long timeMilli = System.currentTimeMillis();
> 		
> 		// Calculate the TimeZone specific offset (including DST)
> 		int timeOffset = timeZone.getOffset(timeMilli)/1000;
> 
> 		return Math.floorMod(secondsInDay + (TownySettings.getNewDayTime() - ((timeMilli/1000) % secondsInDay) - timeOffset), secondsInDay);
> 	}
> ```
> 
> on a near-constant basis when this is used in a tab/scoreboard plugin and it should probably call upon a cached result instead of calculating over and over again. I do realize this would mean that the seconds placeholder is inaccurate.

I can't really think of a solid way to implement a cache system without adding more [dependencies](https://aspects.jcabi.com/annotation-cacheable.html) or using hashmaps<IntSinceLastUpdate, CachedString> to store then compare the result, with it being less than x ml causing it to update. When I attempted the hashmap system, it looked fairly "hacked" together. 

In the rare chance that a server owner's plugin is checking the placeholder every second and is causing problems with their server, their tab/scoreboard plugin should have a cache system in place that can define placehodlers that will only attempt to update a players placeholders every x ticks.

**Comment by LlmDl** [comment]
> 

I have found that most tab/scoreboard/hologram plugins are pinging for placeholders fairly often, not much we can do if a server admin wants to lag their servers with PAPI calls though.

**Comment by LoJoSho** [comment]
> > 
> 
> I have found that most tab/scoreboard/hologram plugins are pinging for placeholders fairly often, not much we can do if a server admin wants to lag their servers with PAPI calls though.

Agreed 👍

**Pull request merged**